source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,433 @@
+#include "source/common/network/happy_eyeballs_connection_impl.h"
+
+#include <vector>
+
+namespace Envoy {
+namespace Network {
+
+HappyEyeballsConnectionImpl::HappyEyeballsConnectionImpl(
+    Event::Dispatcher& dispatcher, const std::vector<Address::InstanceConstSharedPtr>& address_list,
+    Address::InstanceConstSharedPtr source_address, TransportSocketFactory& socket_factory,
+    TransportSocketOptionsConstSharedPtr transport_socket_options,
+    const ConnectionSocket::OptionsSharedPtr options)
+    : dispatcher_(dispatcher), address_list_(address_list), source_address_(source_address),
+      socket_factory_(socket_factory), transport_socket_options_(transport_socket_options),
+      options_(options),
+      next_attempt_timer_(dispatcher_.createTimer([this]() -> void { tryAnotherConnection(); })) {
+  connections_.push_back(createNextConnection());
+}
+
+HappyEyeballsConnectionImpl::~HappyEyeballsConnectionImpl() = default;
+
+void HappyEyeballsConnectionImpl::connect() {
+  ASSERT(!connect_finished_);
+  connections_[0]->connect();
+  maybeScheduleNextAttempt();
+}
+
+void HappyEyeballsConnectionImpl::addWriteFilter(WriteFilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->addWriteFilter(filter);
+}
+
+void HappyEyeballsConnectionImpl::addFilter(FilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->addFilter(filter);
+}
+
+void HappyEyeballsConnectionImpl::addReadFilter(ReadFilterSharedPtr filter) {
+  if (connect_finished_) {
+    connections_[0]->addReadFilter(filter);
+    return;
+  }
+  post_connect_state_.read_filters_.push_back(filter);
+}
+
+void HappyEyeballsConnectionImpl::removeReadFilter(ReadFilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->removeReadFilter(filter);
+}
+
+bool HappyEyeballsConnectionImpl::initializeReadFilters() {
+  ASSERT(connect_finished_);
+  return connections_[0]->initializeReadFilters();
+}
+
+void HappyEyeballsConnectionImpl::addBytesSentCallback(Connection::BytesSentCb cb) {
+  ASSERT(connect_finished_);
+  connections_[0]->addBytesSentCallback(cb);
+}
+
+void HappyEyeballsConnectionImpl::enableHalfClose(bool enabled) {
+  ASSERT(connect_finished_);
+  connections_[0]->enableHalfClose(enabled);
+}
+
+bool HappyEyeballsConnectionImpl::isHalfCloseEnabled() {
+  ASSERT(connect_finished_);
+  return connections_[0]->isHalfCloseEnabled();
+}
+
+std::string HappyEyeballsConnectionImpl::nextProtocol() const {
+  ASSERT(connect_finished_);
+  return connections_[0]->nextProtocol();
+}
+
+void HappyEyeballsConnectionImpl::noDelay(bool enable) {
+  if (!connect_finished_) {
+    per_connection_state_.no_delay_ = enable;
+  }
+  for (auto& connection : connections_) {
+    connection->noDelay(enable);
+  }
+}
+
+void HappyEyeballsConnectionImpl::readDisable(bool disable) {
+  ASSERT(connect_finished_);
+  connections_[0]->readDisable(disable);
```

Comment:
Curious here on why it only acts on the first connection but not all connections in connections_

Reply:
I think because it's not supposed to be called until after connect is finished, so there should only be one connection.